### PR TITLE
Add Quantizable MobilenetV3 architecture for Classification

### PR DESCRIPTION
**Pull request by datumbox**
The pre-trained model was trained:

```
python -m torch.distributed.launch --nproc_per_node=8 --use_env train_quantization.py\
--model='mobilenet_v3_large' --wd 0.00001 --lr 0.001
```
Submitted batch job 35496554 

Validated with:
```
python train_quantization.py --device='cpu' --model='mobilenet_v3_large' --test-only
```

Accuracy metrics (Epoch 89):
`Acc@1 73.004 Acc@5 90.858`

### CODE REVIEW

File: torchvision/models/quantization/mobilenetv3.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        backend = 'qnnpack'
+
+        model.fuse_model()
+        model.qconfig = torch.quantization.get_default_qat_qconfig(backend)
+        torch.quantization.prepare_qat(model, inplace=True)
+
+        if pretrained:
+            _load_weights(arch, model, quant_model_urls.get(arch + '_' + backend, None), progress)
+
+        torch.quantization.convert(model, inplace=True)
```
Comment (fmassa) [inline torchvision/models/quantization/mobilenetv3.py:107]: @raghuramank100 the approach we had to follow here for loading the pre-trained weights for a quantized model is [different from what we did for the other models](https://github.com/pytorch/vision/blob/ad3bbef90707d6495803c59a3ef4910c91c5025e/torchvision/models/quantization/mobilenetv2.py#L77-L93). Would you happen to know why we can't use the previous approach

File: docs/source/models.rst [hunk 2]
```diff
@@ -263,6 +263,53 @@ MNASNet
...
+    # run the model with quantized inputs and weights
+    out = model(torch.rand(1, 3, 224, 224))
+
+We provide pre-trained quantized weights for the following models:
+
+================================  =============  =============
+Model                             Acc@1          Acc@5
+================================  =============  =============
+MobileNet V2                      71.658         90.150
+MobileNet V3 Large                73.004         90.858
```
Comment (fmassa) [inline docs/source/models.rst:304]: @raghuramank100 this is ~ 1 acc@1 point drop compared to the fp32 reference. Would you have any tips on how to make this gap smaller?

**Review by fmassa — APPROVED** [review]
Looks great, thanks a lot!

I have a few questions to @raghuramank100 which are not blocking to get this PR merged, but I would love to get this thoughts on a couple of points.

**Comment by fmassa** [comment]
BTW, test failures seem to be related

File: docs/source/models.rst (hunk 2 above)
Comment (datumbox) [reply docs/source/models.rst:304]: @fmassa The non-quantized version of MobileNet V3 Large uses averaging of checkpoints which I don't do here. That's possibly one of the reasons we get lower accuracy.

**Comment by datumbox** [comment]
@fmassa Thanks for flagging. I fixed the issues, did 1 epoch of training and 1 validation to ensure everything still works fine. The rest of the failing tests are not related to this PR. So we should be OK.

File: torchvision/models/quantization/mobilenetv3.py (hunk 1 above)
Comment (raghuramank100) [reply torchvision/models/quantization/mobilenetv3.py:107]: Can you elaborate on why the other approach didnt work? For QAT, we should load the fp32 model (prior to prepare) and then start training with that.

File: torchvision/models/quantization/mobilenetv3.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    **kwargs: Any
+):
+    model = QuantizableMobileNetV3(inverted_residual_setting, last_channel, block=QuantizableInvertedResidual, **kwargs)
+    _replace_relu(model)
+
+    if quantize:
+        backend = 'qnnpack'
+
+        model.fuse_model()
+        model.qconfig = torch.quantization.get_default_qat_qconfig(backend)
```
Comment (raghuramank100) [inline torchvision/models/quantization/mobilenetv3.py:101]: One suggestion for improvement is to change the qconfig to the following: This configuration uses per-channel quantization for weights with qnnpack, which is now supported.
        qconfig = QConfig(activation=torch.quantization.FakeQuantize.with_args(observer=torch.quantization.MovingAverageMinMaxObserver,
                                                            quant_min=0,
                                                            quant_max=255,
                                                            reduce_range=False),
                          weight= default_per_channel_weight_fake_quant)

File: docs/source/models.rst (hunk 2 above)
Comment (raghuramank100) [reply docs/source/models.rst:304]: If you start with the averaged checkpoint to start quantization aware training, you should get better accuracy as the starting point is better.

Comment (raghuramank100) [reply docs/source/models.rst:304]: Also, one additional hyper-parameter that helps is to turn on QAT in steps: We first turn observers on (i.e collect statistics) and then turn fake-quantization on, and after sometime we turn batch norm off. Currently, in train_quantization, steps 1 and 2 are combined. We have seen that separating them helps with QAT accuracy in some models.

Comment (datumbox) [reply docs/source/models.rst:304]: > If you start with the averaged checkpoint to start quantization aware training, you should get better accuracy as the starting point is better.

We indeed start from an averaged checkpoint but that's not what I mean here. I'm referring to the post-training averaging step which is missing.

> We first turn observers on (i.e collect statistics) and then turn fake-quantization on.

That's worth integrating on the new quant training script.

I believe key reason why the accuracy is lagging is because the quant training script does not currently support all the enhancements made on the classification training script. These enhancements (Multiple restarts, Optimizer tuning, Data augmentation, model averaging at the end etc) helped me push the accuracy by 2 points.

File: torchvision/models/quantization/mobilenetv3.py (hunk 1 above)
Comment (datumbox) [reply torchvision/models/quantization/mobilenetv3.py:107]: > For QAT, we should load the fp32 model (prior to prepare) and then start training with that.

We did not face any issues during the training process and what you describe is the approach we used for training the model.

> Can you elaborate on why the other approach didnt work?

After the training  was completed, we tried to load the weights of the quantized model (key "model_eval" of the checkpoint) to do inference. Unfortunately doing so leads to extremely low accuracy (less than 1%). We are certain that the weights are loaded on the model.

As a workaround, we opted for loading the weights of the QAT model (key "model"` of the checkpoint) and then convert it. This works fine and gives the same inference accuracy as observed during training. 

We are trying to understand what could be the reason for this behaviour. I can provide a demo script if that helps.
